core/src/main/java/org/apache/iceberg/BaseRewriteFiles.java

```diff
@@ -85,4 +90,18 @@ public RewriteFiles rewriteFiles(Set<DataFile> dataFilesToDelete, Set<DeleteFile
 
     return this;
   }
+
+  @Override
+  public RewriteFiles validateFromSnapshot(long snapshotId) {
+    this.startingSnapshotId = snapshotId;
+    return this;
+  }
+
+  @Override
+  protected void validate(TableMetadata base) {
+    if (replacedDataFiles.size() > 0) {
+      // if there are replaced data files, there cannot be any new row-level deletes for those data files
+      validateNoNewDeletesForDataFiles(base, startingSnapshotId, replacedDataFiles);
```

Comment:
I think we should also handle the case : 

* t1:  table has FILE_A and EQ_DELETE_FILE_B;
* t2:  RewriteFiles action to convert the EQ_DELETE_FILE_B to POS_DELETE_FILE_C which hold the file_id and offset from FILE_A,  just start the txn but not commit;
* t3:  DeleteFiles action committed the txn which deletes  the FILE_A;
* t4:  Try to commit the RewriteFiles action. 

At timestamp `t4`,  the RewriteFiles txn should be aborted because the data file FILE_A which is relied by POS_DELETE_FILE_C has been deleted. 

Do we plan to add this validation in the following PR, or did not consider this case because we currently don't have introduced any RewriteFiles action that converting equality delete files into positional delete files ?

Reply:
In my thought,   the data conflicts between REPLACE operation and APPEND/OVERWRITE/DELETE operations is because of the dependency relationship from positional delete files to data files. Then there will be two cases:

1.  The REPLACE operation remove the data files that was relied by other committed APPEND/OVERWRITE/DELTE  operations ( among the three operations,  currently only the RowDelta will introduce new positional delete files which reference to committed data files).  I think we handle this case perfectly in this PR.
2. The APPEND/OVERWRITE/DELETE operations remove the data files that was relied by other committed REPLACE operation.  The above [comment](https://github.com/apache/iceberg/pull/2865/files#r677072072) is an example. 

As a final solution,  I think we should handle both the cases perfectly, so that we won't encounter any unexpected interruption when reading the iceberg table.

Reply:
@openinx, I opened a draft PR, https://github.com/apache/iceberg/pull/2892, for to fix this, but I'm not sure that we actually want to merge it.

Originally, I copied the `validateDataFilesExist` implementation from `RowDelta` and added tests. That works great and is a fairly simple commit if we want it. But as I was testing it, I added cases for delete, overwrite, rewrite, and a transactional delete and append to overwrite.

Then I realized that not all of those cases apply. For example, if I delete a file that has row deltas, that's perfectly okay. The operation actually deletes all rows that weren't already deleted and any dangling positional deletes will just never get applied. Similarly, an overwrite from, for example, a MERGE INTO operation that replaces a file will need to first read that file and then build its replacement. The read must merge the existing deletes, so the rows will already be removed from the replacement file. Then the deletes in that case are also fine to leave dangling even if they are rewritten into a new file.

I ended up leaving a validation that the data files are not rewritten, but I also think that may not be necessary because I don't think it would be reasonable to rewrite a data file without also applying the deletes. If we build a compaction operation we will probably apply deletes during compaction. We could also build one that compacts just data files, but in order for that operation to be correct it would also need to rewrite delete files (that are pointing to the compacted data files) or use an older sequence number with equality deletes. If it rewrote delete files, then the delete rewrite would fail. If it changed sequence numbers, then we would have a lot more validation to do when building that plan.

After thinking about it, I don't think that we actually need to handle any cases where a compaction and delete file rewrite are concurrent, at least not from the delete file rewrite side. Any compaction should handle deletes as part of the compaction.

Reply:
> We could also build one that compacts just data files, but in order for that operation to be correct it would also need to rewrite delete files (that are pointing to the compacted data files) or use an older sequence number with equality deletes

You mean people will need to know whether the  data-files-only compaction will break the existing data set ? This puts too much burden on users ?  I think we still need to provide the validation because we developers are sure that those cases will break the data correctness ...

Reply:
> You mean people will need to know whether the data-files-only compaction will break the existing data set ? This puts too much burden on users ? I think we still need to provide the validation because we developers are sure that those cases will break the data correctness ...

The contract of the `RewriteFiles` operation is that it does not change the table data. That is explicitly stated in the `rewriteFiles` configuration method's javadoc:

> Add a rewrite that replaces one set of files with another set that **contains the same data**.

I think that is a reasonable requirement for the operation. Compacting only data files without accounting for deletes would change the table data, even without a concurrent delete file rewrite. For example, if I have

| `file-a.parquet` |
| -- |
| `1, 'x'` |
| `2, 'y'` |

| `file-b.parquet` |
| -- |
| `3, 'z'` |

| `file-a-deletes.parquet` |
| -- |
| `'file-a.parquet', 0` |

Compacting `file-a.parquet` and `file-b.parquet` together would undelete row `id=1`. There is no way for Iceberg to check whether the rewrite ignored `file-a-deletes.parquet` or applied the deletes. The correctness problem is entirely in the compaction operation.

We have to trust the `RewriteFiles` caller to do the right thing and preserve correctness.

That means either the caller must do one of the following:
1. Remove the deleted row
2. Rewrite `file-a-deletes.parquet` to point the delete at the new data file
3. Write the new data file with a compatible sequence number (only valid if the deletes are equality deletes so they would still apply)

If the caller uses option 1, then it doesn't matter to concurrent operations that are rewriting the delete. It becomes a dangling delete that doesn't need to be applied. If the caller uses option 2, then there is a new file tracking the delete that would not be part of a concurrent rewrite operation; a concurrent rewrite may contain a dangling delete.

The only case we would need to worry about is option 3, and only if a concurrent delete rewrites an equality delete as a positional delete. To be careful, we could merge #2892 to be able to validate this case. But I'm skeptical that altering the sequence number of the compacted file is a good idea. It may not even be possible if the sequence number of a delete that must be applied is between the sequence numbers of the starting data files.

As I said, I'm okay not merging the validation for option 3 because I think it is unlikely that this is going to be a viable strategy. But I'm fine merging #2892 if you think it is needed.